### ISSUE TITLE
Some people might be disrupted if Eastwood auto-adds itself as a dependency

### ISSUE THREAD

**Issue by jafingerhut** [jonase/eastwood#286]
I still don't know the Leiningen internals well enough to know all the tradeoffs and issues here, but wanted to capture this idea in an issue, at least, since a different Clojure tool auto-adds itself as a dependency to all Leiningen projects, even if it is only in your `~/.lein/profiles.clj` as a plugin.

If Eastwood decides to go this way, probably good to at least document it semi-prominently in the README.  Maybe there is a clean way to avoid doing that, though.

Copy-and-paste of some 2018-Sep-26 discussion on Slack:

lvh [11:07 AM]
I have a project where dependencies are ostensibly being included from ~/.lein/profiles.clj into releases and I can't for the life of me figure out why: https://github.com/cloverage/cloverage/issues/228#issuecomment-424810468

hiredman [11:25 AM]
it looks like ultra adds itself as a project dep
https://github.com/venantius/ultra/blob/master/src/ultra/plugin.clj#L81-L95

jaawerth [11:57 AM]
I like ultra, but that is sneaky

**Comment by jafingerhut** [comment]
A few follow-up comments in Slack discussion:

seancorfield [12:24 PM]
@andy.fingerhut I'm curious: why would Eastwood do that? What does it gain by inserting itself like that?

andy.fingerhut [12:25 PM]
All currently released versions of Eastwood do not do this.

seancorfield [12:25 PM]
(and it seems like venantius/ultra crops up regularly enough that it pretty much needs to be an FAQ in the Clojure world about "why are my dependencies broken?" :slightly_smiling_face: )

andy.fingerhut [12:26 PM]
It was a recently committed change, not yet part of any release, that I believe has such a change.

**Comment by jafingerhut** [comment]
Probably the last of the conversation on this topic copied below.  It looks like I may have been imagining that what ultra does, and what Eastwood does, are more similar than they are in reality, so this issue might amount to "no changes to Eastwood are desirable".

seancorfield [12:41 PM]
Ah, I see what it's doing now. Previously, it used to add itself as a dependency if it wasn't already specified (via `add-if-missing` from `leinjacker`). Now, it looks for a plugin dependency if present and uses that to add itself as a (dev) dependency. And it only seems to do it while Eastwood is running inside Leiningen (`eval-in-project`).
So it doesn't look like the behavior has really changed -- it's just dropped support for Leiningen 1.x and relies on core Leiningen stuff in 2.x to run itself now?
I think the difference with Ultra is that it injects itself into your deps in all situations?

andy.fingerhut [12:43 PM]
I have not dug into how ultra modifies things, or these recent Eastwood changes, enough to understand them fully.  I raised the issue because I was pattern matching on "ultra behavior X causes some people problems.   Hmm, that sounds similar to a recent Eastwood change that maybe shouldn't be released if it would cause the same problems."
If you have more details that might help make good decisions there, and have the time and interest to share your knowledge, that would be great.

seancorfield [12:46 PM]
The above is based on 1) just now reading the entire commit list for Eastwood since the last release and 2) repeated complaints from people who've tripped over Ultra's behavior when they're not aware they're explicitly "using" it for anything... so I'm not sure I'm any more qualified than the next person )

slipset [12:49 PM]
@seancorfield your reading of the dependency injection code in Eastwood very much resembles the intent of it.

andy.fingerhut [12:50 PM]
@slipset Sorry if I am raising alarms for no good reason.  My recent comment above gives what could loosely be called my thinking process here.

slipset [12:50 PM]
Basically, as far as I understand, if you have Eastwood only listed as a plug-in, lein needs it as a dep to actually run it.
@andy.fingerhut no problem. I’m just grateful for more eyes on this.

**Comment by jafingerhut** [comment]
Closing this issue, as I think I jumped the gun of there being a possible problem, when in fact it seems like things are still OK here.
